### PR TITLE
Fix invalid computations in VAT

### LINKED ISSUE #120: Invalid values in KL divergence computation of VAT
**Describe the bug**
This issue was discovered during the development for #75. Because the model predictions / logits can be negative, the computation of the KL divergence in virtual adversarial method can yield invalid values (`nan` and `inf`), preventing the method from producing meaningful results.

**System information (please complete the following information):**
 - OS: Ubuntu
 - Python version: 3.6
 - ART version or commit number: 3d618e3

### PR DESCRIPTION
**Pull request by ririnicolae**
# Description

Fix issue with numerical computation in VAT. As described in the original issue (#120), one of the root causes was related to invalid values encountered in the computation of KL divergence. This was coming from negative values predicted in classifiers. Moreover, an additional root cause was present in the computation of L_2 data normalization in VAT. This PR fixes both issues. As a consequence, VAT can be used on both images and feature vectors without restrictions and warnings.

Fixes #120, #157

## Type of change

Please check all relevant options.

- [X] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [X] Existing unit tests
- [X] Uncommented unit tests that were previously failing

**Test Configuration**:
- OS: Ubuntu
- Python version: 3.6
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by minhitbk** [comment]
I am a bit confused on the _rescale function. I had a quick look at the paper and the original code of VAT of the authors, I dont see that rescale concept (or maybe I missed it). I think VAT is designed to work with probability output, so users need to guarantee that their model should produce probability. In this case, we don't need to rescale or change the current implementation of VAT (except for the small bug in the normalize function). Instead, we only need to raise error to users if they input a non-probability output model, and fix bugs in the unittests.

In case the concept of rescale is available somewhere in VAT but I missed, then the code looks well to me and ready to merge.

**Comment by ririnicolae** [comment]
@minhitbk You are right, the `_rescale` function is not part of the original VAT paper. As you mentioned, the method was conceived to only be applied on top of softmax. However, ART now handles many more types of models than DNNs with softmax. Introducing the `_rescale` function is an extension that allows to generalize to these models. I thought this would be a good idea, considering we have adapted other methods in the past for better performance or extended use cases.
